### DISCUSSION TITLE
Device Authorization grant doesn't complete, via Azure?

### DISCUSSION THREAD

**Discussion by tashian** [smallstep/certificates#1176] [Q&A] [answered]
### Steps to Reproduce

In the Azure Active Directory portal:
- Create a new App Registration in Azure Active Directory, using the default settings. Note the client ID on the main page for the App Registration.
- Create a Client Secret, using the default settings. Note the secret.
- Finally, choose "Endpoints" to find the URL for the OIDC configuration endpoint

Now, create an OIDC provisioner with these values.

Next, try to get a certificate, using console mode:

```
STEPDEBUG=1 step ca certificate carl@smallstep.com carl.crt carl.key --console --provisioner azure
✔ Provisioner: azure (OIDC) [client: df32d272-7dda-4c8a-8d40-cc89218c2a87]
Visit https://microsoft.com/devicelogin and enter the code: (press 'ENTER' to open default browser)
EWPU7M86B
```

When you visit https://microsoft.com/devicelogin and enter the code, the OAuth flow will complete successfully.
But the flow in the CLI will never complete.

### Your Environment

* OS - darwin
* `step` CLI Version - Smallstep CLI/0.21.0-29-g3586b16

### Expected Behavior

I expected to get a certificate after the flow completed successfully.

### Actual Behavior

Once the flow completes successfully, nothing happens in the terminal.

### Additional Context

Running with `STEPDEBUG=1` provides no additional information.

**Comment by maraino** [comment, marked as answer]
Hi @tashian, Azure returns the following
```json
{
  "error": "invalid_client",
  "error_description": "AADSTS7000218: The request body must contain the following parameter: client_assertion or client_secret....",
  "error_codes": [7000218],
  "timestamp": "xxx",
  "trace_id": "yyy",
  "correlation_id": "zzz",
  "error_uri": "https://login.microsoftonline.com/error?code=7000218"
}
```

Looking for AADSTS7000218 this [post](https://stackoverflow.com/questions/67573253/get-msaltoken-error-aadsts7000218-the-request-body-must-contain-the-following-p) in stackoverflow suggest editing the manifest of the app and setting `"allowPublicClient": true`. After this my client was fixed.
